pkg/shared/accountmgmtutil/ams.go

```diff
@@ -68,144 +70,180 @@ func fetchOrgQuotaCost(ctx context.Context, conn connection.Connection) (*amscli
 	}
 
-	return account.Organization.GetId(), nil
+	return &orgQuota.TrialQuotas[0], nil
 }
 
-func GetValidMarketplaceAcctIDs(ctx context.Context, connectionFunc factory.ConnectionFunc, marketplace string) (marketplaceAcctIDs []string, err error) {
-
-	conn, err := connectionFunc(connection.DefaultConfigSkipMasAuth)
-	if err != nil {
-		return nil, err
+func getMarketplaceQuota(f *factory.Factory, marketplaceQuotas []QuotaSpec, marketplace MarketplaceInfo) (*QuotaSpec, error) {
+	if len(marketplaceQuotas) == 1 {
+		if marketplace.Provider != "" || marketplace.CloudAccountID != "" {
+			marketplaceQuota, err := pickMarketplaceQuota(f, marketplaceQuotas, marketplace)
+			if err != nil {
+				return nil, err
+			}
+			return marketplaceQuota, nil
+		}
+		return &marketplaceQuotas[0], nil
+	} else if len(marketplaceQuotas) > 1 && marketplace.Provider == "" && marketplace.CloudAccountID == "" {
+		return nil, f.Localizer.MustLocalizeError("kafka.create.quota.error.multipleMarketplaceQuotas")
 	}
 
-	quotaCostGet, err := fetchOrgQuotaCost(ctx, conn)
+	marketplaceQuota, err := pickMarketplaceQuota(f, marketplaceQuotas, marketplace)
 	if err != nil {
 		return nil, err
 	}
+	return marketplaceQuota, nil
 
-	for _, quota := range quotaCostGet.GetItems() {
-		if len(quota.GetCloudAccounts()) > 0 {
-			for _, cloudAccount := range quota.GetCloudAccounts() {
-				if marketplace != "" {
-					if cloudAccount.GetCloudProviderId() == marketplace {
-						marketplaceAcctIDs = append(marketplaceAcctIDs, cloudAccount.GetCloudAccountId())
-					}
-				} else {
-					marketplaceAcctIDs = append(marketplaceAcctIDs, cloudAccount.GetCloudAccountId())
+}
+
+func pickMarketplaceQuota(f *factory.Factory, marketplaceQuotas []QuotaSpec, marketplace MarketplaceInfo) (*QuotaSpec, error) {
+
+	matchedQuotas := []QuotaSpec{}
+
+	for _, quota := range marketplaceQuotas {
+		cloudAccounts := *quota.CloudAccounts
+		for _, cloudAccount := range cloudAccounts {
+			if marketplace.Provider != "" && marketplace.CloudAccountID != "" {
+				if *cloudAccount.CloudProviderId == marketplace.Provider && *cloudAccount.CloudAccountId == marketplace.CloudAccountID {
+					matchedQuotas = append(matchedQuotas, quota)
+				}
+			} else if marketplace.Provider != "" || marketplace.CloudAccountID != "" {
+				if *cloudAccount.CloudProviderId == marketplace.Provider || *cloudAccount.CloudAccountId == marketplace.CloudAccountID {
+					matchedQuotas = append(matchedQuotas, quota)
 				}
 			}
 		}
 	}
 
-	return unique(marketplaceAcctIDs), err
+	if len(matchedQuotas) == 0 {
+		return nil, f.Localizer.MustLocalizeError("kafka.create.quota.error.cloudAccountNotFound")
+	}
+
+	return &matchedQuotas[0], nil
 }
 
-func GetValidMarketplaces(ctx context.Context, connectionFunc factory.ConnectionFunc) (marketplaces []string, err error) {
+func pickCloudAccount(f *factory.Factory, cloudAccounts *[]amsclient.CloudAccount, market MarketplaceInfo) (*[]amsclient.CloudAccount, error) {
```

Comment:
Should this method return single object instead of array. It does `pick` cloud account right?

Reply:
Used array to make it easy to use with quotapsec object.